### DISCUSSION TITLE
关于 SmartDNS 配置的讨论

### DISCUSSION THREAD

**Discussion by jxph1123** [Openwrt-Passwall/openwrt-passwall#2417] [General]
关于luci-smartdns-new-version，目前passwall国内组在smartdns内自定义,只需要填写分组名，而国外组是passwall写好配置自动生成。
建议国外组也在smartdns内自定义，和国内组一样只需要填写分组名。有几个好处：
1.在smartdns内设置好后，无需再到passwall内设置国外DNS服务器。
2.用samrtdns配置国外组参数，配置相对灵活，不像现在国外组参数写死了。
欢迎大家讨论。

**Comment by uingei** [comment]
我还是继续老模式smartdns两个组，一个端口给dnsmasq一个端口给passwall，不知道有啥问题，用起来貌似还行

**Reply by jxph1123** [reply]
我感觉这样也行啊,不清楚有啥弊端

**Reply by uingei** [reply]
试了luci-smartdns-new-version感觉比原来走两端口好

**Comment by zqp561022** [comment]
现在smartdns也可以设置自己的国外组例如“us"组，核心是配置好代理服务器“socks5://127.0.0.1:1070”、国外组配置上游服务器时"从默认服务器组排除"、"使用代理连接上游 DNS 服务器"，然后用这个国外组在smartdns的域名规则走IPSet黑名单配合passwall分流，逻辑也清晰。其实核心只要是域名规则分流后的国外域名查询走代理查就行，至于界面怎么走都还好。

**Reply by handsomeji** [reply]
Yes, it is completely correct.

**Reply by jxph1123** [reply]
> 现在smartdns也可以设置自己的国外组例如“us"组，核心是配置好代理服务器“socks5://127.0.0.1:1070”、国外组配置上游服务器时"从默认服务器组排除"、"使用代理连接上游 DNS 服务器"，然后用这个国外组在smartdns的域名规则走IPSet黑名单配合passwall分流，逻辑也清晰。其实核心只要是域名规则分流后的国外域名查询走代理查就行，至于界面怎么走都还好。

我感觉你这样配置更麻烦哇,能发下你的配置吗,谢谢~~

**Reply by wtfr-dot** [reply]
> 现在smartdns也可以设置自己的国外组例如“us"组，核心是配置好代理服务器“socks5://127.0.0.1:1070”、国外组配置上游服务器时"从默认服务器组排除"、"使用代理连接上游 DNS 服务器"，然后用这个国外组在smartdns的域名规则走IPSet黑名单配合passwall分流，逻辑也清晰。其实核心只要是域名规则分流后的国外域名查询走代理查就行，至于界面怎么走都还好。

逻辑是没错，但是有两个问题：1、如果不用目前passwall预设服务器，则需要采用127.0.0.1:port这种方式与smartdns通讯，但是smartdns会把127.0.0.1:port当成一个服务器组加载，造成了重复，逻辑上不清，2、最主要的是如果你自己在smartdns那里设置了ipset规则，但passwall的分流不是依据简单的黑白名单，而且规则是经常更新变化的，但smartdns目前没有简单的规则转换和更新机制，这样造成跟passwall的规则不同步和不一致，有些时候分流可能会出问题。
综上，问题1可以通过通过修改代码解决，我之前就是将127.0.0.1注释掉解决的，但现在改动较大，改起来很麻烦，也没去琢磨了。问题2实际上不需要自己在smartdns那里再自设ipset规则了，passwall已经做好分类转换，确定分流规则后会自动发往相应的服务器组查询。
目前的使用方案确实不太灵活，smartdns也经常在增加新的参数，但这样就用不上了，虽然不知道是否真的要用，但目前的使用方案对我来说也够了。

**Reply by wtfr-dot** [reply]
> > > 现在 smartdns 也可以设置自己的国外组例如 “us"组，核心是配置好代理服务器 “socks5://127.0.0.1:1070”、国外组配置上游服务器时" 从默认服务器组排除 "、" 使用代理连接上游 DNS 服务器 "，然后用这个国外组在 smartdns 的域名规则走 IPSet 黑名单配合 passwall 分流，逻辑也清晰。其实核心只要是域名规则分流后的国外域名查询走代理查就行，至于界面怎么走都还好。
> > 
> > 
> > 我感觉你这样配置更麻烦哇，能发下你的配置吗，谢谢～～
> > 个人配置思路，仅供参考哈
> > 一、	配置思路
> > 网络请求将会被封装转发给远端代理服务器、会在远端代理服务器进行 DNS 解析，smartdns 作为主 dns，可以在第二 DNS 服务器不用的情况下分流。对于有特殊需求的域名（us）走分流组；其他非特殊需求的（cn）走测速。对应的，只要设置好 passwall 的 ipset 名称和走透明代理的列表即可（这里分流是黑名单模式），这样可以将有需求的域名走代理，其他域名测速直连。
> > 二、配置lan口（我是旁路由模式，主路由模式请自行理解）
> > 【常规设置】-网关指向主路由地址：192.168.0.1（例），【高级设置】使用自定义DNS服务器地址，指向op的本地IP：127.0.0.1，【DHCP】-忽略此接口，【动态DHCP】关闭，关闭DHCP服务，禁用IPV6相关，由主路由负责提供DHCP。
> > 三、smartdns设置
> > 1、配置第一组smartdns服务端口及设置为dnsmasq上游
> > 【基本设置】-本地端口：6053（可自定义），自动设置为Dnsmasq上游勾选，后面启动成功后可以到DHCP处看下Dnsmasq转发是自动修改为转发到127.0.0.1#6053。
> > 2、配置国内组上游服务器
> > 可配置本地运营商的UDP+阿里腾讯的DOH+DOT，【服务器组】：cn（可自定义），不勾选从默认组排除。smartdns 里的上游 dns 服务器只要不设置排除默认组，它们都是默认组的成员。
> > 3、配置代理服务器
> > socks5://127.0.0.1:1070
> > 这个端口号是 passwall 配置中的 “基本设置 --> 主要 --> TCP 节点 Socks 监听端口” （版本过低有可能是没有这一项内容的）。
> > 4、配置国外组上游服务器
> > 设置【国外】的 DOT，DOH 等服务器，【服务器组】： us（可自定义）。勾选‘从默认组中排除’，勾选‘使用代理’，这样查询这些国外DNS服务器都是通过passwall代理走转发程序，目的国外域名通过代理查询，DNS 结果与机场适配更好，而不是查询的你自己本地最快。‘从默认组中排除’或者是-exclude-default-group 告诉 smartdns，非特殊域名查询时，不要使用这个 DNS 服务器。
> > 5、配置域名列表文件自动下载
> > 【下载文件】-文件名：proxy-list.txt；URL：[https://cdn.jsdelivr.net/gh/Loyalsoldier/v2ray-rules-dat@release/proxy-list.txt；文件类型：域名列表。](https://cdn.jsdelivr.net/gh/Loyalsoldier/v2ray-rules-dat@release/proxy-list.txt%EF%BC%9B%E6%96%87%E4%BB%B6%E7%B1%BB%E5%9E%8B%EF%BC%9A%E5%9F%9F%E5%90%8D%E5%88%97%E8%A1%A8%E3%80%82)
> > 【上传域名列表文件】路径填写：/etc/smartdns/domain-set/proxy-list.txt
> > 更新列表文件，启用自动更新。正确填写并更新后可以看到文件大小，“上传的文件” (约500多 KB)。
> > 【域名规则】-域名分流设置-【服务器组】：us；跳过测速，停用 IPV6 地址解析；IPSet 名称：gfwlist；【域名列表文件】路径填写：/etc/smartdns/domain-set/proxy-list.txt；正确填写并更新后可以看到文件大小，“上传的文件” (约500多 KB)。
> > 对于临时个别要分流的域名，可以在域名列表文本框中输入。
> > 原理上，smartdns 和转发软件的结合点，为 ipset 名称或 nftset 名称，那两个参数和转发软件匹配就可以做到按需转发。
> > 6、高级设置
> > 测速模式：ping,tcp80,tcp443;响应模式：最快ping；停用 IPV6 地址解析；域名预加载和缓存过期服务根据自身需要选择；
> > 【无速度时 IPSet 名称】：blacklist
> > 将测速失败的 IP 地址，加入 IPSet，再由相关 IP 规则转发。
> > 四、passwall设置(smartdns分支）
> > 1、dns设置
> > dns分流：smartdns，国内分组名：cn，远程 DNS：随便+
> > 2、模式设置
> > TCP 默认代理模式：GFW列表
> > UDP 默认代理模式：GFW列表
> > 路由器本机 TCP 代理模式、路由器本机 UDP 代理模式与默认相同
> > 其他passwall的设置该咋设置咋设置，这个流程是利用smartdns+passwall走黑名单模式，原理搞懂了也可以在smartdns那里用白名单域名列表配合国内组+passwall走中国列表以外模式。

你这样做把简单的问题复杂话了，白费了passwall的努力，这个配置适合搭配其它代理软件，目前passwall应该是与smartdns配合最紧密的代理软件了

**Comment by jxph1123** [comment]
就是想简化,我的建议是passwall最简单设置,国内国外分别设置到smartdns的分组名,其他的交给smartdns设置就行了.

**Reply by wtfr-dot** [reply]
目前的方案对小白最友好，对爱折腾的人来说不够灵活，@xiaorouji大佬是否可以考虑这个提议

**Reply by jxph1123** [reply]
我觉得Chnlist、gfwlist等规则转换为smartdns能识别的规则这种复杂的活可以passwall后台干,一些基础的设置,每个分组比如是否过滤IPV6，测速、缓存这些简单的设置可以交给smartdns。

**Comment by jxph1123** [comment]
现在如果使用Dnsmasq设置`当使用中国列表外时的默认DNS选择为"远程"时`,会产生几万个规则,需要Dnsmasq来处理.明显延时会多几十ms,而这个功能在smartdns里面,只需要默认组简单设置为国外DNS就可以实现.虽然smartdns也会处理passwall转换的几万个规则,但是smartdns处理效率明显高于Dnsmasq,基本无延时.
所以smartdns配合passwall,简直绝配.所以建议passwall里面再简化下,直接两个分组,一个国内,一个国外.passwall只需要后台给smartdns转换规则,其余的DNS相关设置就交给专业的smartdns.比如过滤IPV6，测速、缓存、自定义域名等设置。

**Comment by wtfr-dot** [comment]
把这个议题顶一下，目前的这种配置方式确实对小白很友好，但是默认配置下smartdns解析的ip地址在远程vps实际上是被还原成域名的，远程解析的ip地址跟smartdns解析的ip并不一定一致，造成smartdns解析的结果没多大作用，我看奈飞就很明显，奈飞连接的ip地址离vps比较远，然后我将routeonly打开后，远程vps直接接受本地smartdns解析的ip地址，奈飞连接的ip地址就变成了vps所在的城市，但是这样一来，本地的系统会有点小问题，passwall这里域名解析策略我使用的IPIfNomatch，兜底的默认是代理，采用默认的规则下，会有很多域名不在黑白名单内，smartdns那里默认是国内解析组，这样有部分不在黑白名单内又是国外域名会被smartdns的国内组解析而被虚假回应，造成连接失败，如果采用默认规则下载url，得到的gfwlist、chnlist和chnroute结果会比较少，漏掉的域名会比较多，只有多添加几个url，这样会有所缓解，但还是有部分我想访问的域名漏掉了，造成访问失败，为此我只有将smartdns的国内组排除在默认组外，另外新增一个国外dns服务器，将其设为默认组内，再使用代理方式进行解析来解决漏网的域名解析，这样结果倒是正常了，但smartdns却增加了额外的一个dns服务器，形成了总共三个服务器组，多出的这个dns服务器实际上有点多余了，但目前的这种配置又必须设，所以还是建议远程dns也采用国内分组模式进行调用，由smartdns那里设置分组，配合这边的规则进行解析，这样灵活性和逻辑性都更好。

**Reply by jxph1123** [reply]
我目前也是这样，但是我觉得至少还得3个组，默认组、国内组、国外组，毕竟GFW规则、gfwlist列表等不可能区分全部域名，所以总有漏的，这个漏的就得走默认组，具体默认组你要设置国内DNS服务器还是国外DNS，就由你自己选择了。当然，设置国外可以防DNS泄漏，但是对部分网站访问有点影响，就看你怎么取舍了。

**Reply by wtfr-dot** [reply]
如果由smartdns设置服务器组就两个组就可以了，默认组和国外组可以合并，或者默认组和国内组合并，就可以做到漏掉的走代理或者直连，但现在如果漏掉的走代理就得多设一个组，如果走直连则不需要

**Reply by lxb158** [reply]
我咋看不懂啊，上有我有二组名称 一组china，一组office（配置排除默认及走代理），china组要不要配置那里改成默认，不然我没看到有默认组啊

**Comment by smallprogram** [comment]
### 最简单的方案
#### smartdns配置
- 默认DNS，设置非53端口，例如50053，配置dnsmasq 53转发给127.0.0.1:50053，smartdns默认自动配置了，给一堆国内知名dns包含在默认分组中
- 第二DNS，例如配置60053，给一堆国外DNS，配置为排除默认分组，分组名给GFW或者你自己喜欢的名字

#### passwall配置
- DNS分流 chinaDNS-NG，UDP到127.0.0.1:50053
- 远程DNS UDP到127.0.0.1:60053
- DNS重定向开与不开均可
- 停用HTTPS记录解析开与不开均可

#### 唯一缺点
- smartdns日志ip源全是127.0.0.1

#### 注意点
- smartdns绝对不能替换掉dnsmasq的53端口，必须用转发，如果替换，passwall的防火墙规则必定死循环
